enterprise/server/content_addressable_storage_server_proxy/content_addressable_storage_server_proxy.go

```diff
@@ -6,58 +6,131 @@ import (
-	casServer, err := NewCASServerProxy(env)
+	casServer, err := New(env)
 	if err != nil {
 		return status.InternalErrorf("Error initializing ContentAddressableStorageServerProxy: %s", err)
 	}
 	env.SetCASServer(casServer)
 	return nil
 }
 
-func NewCASServerProxy(env environment.Env) (*CASServerProxy, error) {
-	localCache := env.GetCache()
-	if localCache == nil {
-		return nil, fmt.Errorf("A cache is required to enable the ContentAddressableStorageServerProxy")
+func New(env environment.Env) (*CASServerProxy, error) {
+	local := env.GetLocalCASClient()
+	if local == nil {
+		return nil, fmt.Errorf("A local ContentAddressableStorageClient is required to enable the ContentAddressableStorageServerProxy")
 	}
-	remoteCache := env.GetContentAddressableStorageClient()
-	if remoteCache == nil {
-		return nil, fmt.Errorf("A ContentAddressableStorageClient is required to enable the ContentAddressableStorageServerProxy")
+	remote := env.GetContentAddressableStorageClient()
+	if remote == nil {
+		return nil, fmt.Errorf("A remote ContentAddressableStorageClient is required to enable the ContentAddressableStorageServerProxy")
 	}
 	return &CASServerProxy{
-		env:         env,
-		localCache:  localCache,
-		remoteCache: remoteCache,
+		env:    env,
+		local:  local,
+		remote: remote,
 	}, nil
 }
 
+// TODO(iain): update remote atimes.
 func (s *CASServerProxy) FindMissingBlobs(ctx context.Context, req *repb.FindMissingBlobsRequest) (*repb.FindMissingBlobsResponse, error) {
-	return s.remoteCache.FindMissingBlobs(ctx, req)
+	resp, err := s.local.FindMissingBlobs(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.MissingBlobDigests) == 0 {
+		return resp, nil
+	}
+	remoteReq := repb.FindMissingBlobsRequest{
+		InstanceName:   req.InstanceName,
+		BlobDigests:    resp.MissingBlobDigests,
+		DigestFunction: req.DigestFunction,
+	}
+	return s.remote.FindMissingBlobs(ctx, &remoteReq)
 }
 
 func (s *CASServerProxy) BatchUpdateBlobs(ctx context.Context, req *repb.BatchUpdateBlobsRequest) (*repb.BatchUpdateBlobsResponse, error) {
-	return s.remoteCache.BatchUpdateBlobs(ctx, req)
+	_, err := s.local.BatchUpdateBlobs(context.Background(), req)
+	if err != nil {
+		log.Warningf("Local BatchUpdateBlobs error: %s", err)
+	}
+	return s.remote.BatchUpdateBlobs(ctx, req)
 }
 
 func (s *CASServerProxy) BatchReadBlobs(ctx context.Context, req *repb.BatchReadBlobsRequest) (*repb.BatchReadBlobsResponse, error) {
-	return s.remoteCache.BatchReadBlobs(ctx, req)
+	mergedResp := repb.BatchReadBlobsResponse{}
+	mergedDigests := []*repb.Digest{}
+	localResp, err := s.local.BatchReadBlobs(ctx, req)
+	if err != nil {
+		return s.batchReadBlobsRemote(ctx, req)
+	}
+	for _, resp := range localResp.Responses {
+		if resp.Status.Code == int32(codes.OK) {
+			mergedResp.Responses = append(mergedResp.Responses, resp)
+			mergedDigests = append(mergedDigests, resp.Digest)
+		}
+	}
+	if len(mergedResp.Responses) == len(req.Digests) {
+		return &mergedResp, nil
+	}
```

Comment:
I think it is legal (in the protocol) for a client to send the same digest twice. If that happens, will local/remote BatchReadBlobs return both or dedupe?

If it returns both, this code is correct. If it dedupes, then I think you have an issue where you'd fall through and digests might be an empty list?

Reply:
This was actually working due to a bug in `digest.Diff` that I fixed in https://github.com/buildbuddy-io/buildbuddy/commit/d8461cff320849f1f9dc8aef322a15a525f76f10. I added a test, rebased, and fixed the bug.